rsconnect/api.py

```diff
@@ -485,17 +608,61 @@ def deploy_bundle(
         bundle: IO = None,
         env_vars=None,
     ):
-        result = self.client.deploy(
-            app_id or self.get("app_id"),
-            deployment_name or self.get("deployment_name"),
-            title or self.get("title"),
-            title_is_default or self.get("title_is_default"),
-            bundle or self.get("bundle"),
-            env_vars or self.get("env_vars"),
-        )
-        self.connect_server.handle_bad_response(result)
-        self.state["deployed_info"] = result
-        return self
+        app_id = app_id or self.get("app_id")
+        deployment_name = deployment_name or self.get("deployment_name")
+        title = title or self.get("title")
+        title_is_default = title_is_default or self.get("title_is_default")
+        bundle = bundle or self.get("bundle")
+        env_vars = env_vars or self.get("env_vars")
+
+        if isinstance(self.remote_server, RSConnectServer):
+            result = self.client.deploy(
+                app_id,
+                deployment_name,
+                title,
+                title_is_default,
+                bundle,
+                env_vars,
+            )
+            self.remote_server.handle_bad_response(result)
+            self.state["deployed_info"] = result
+            return self
+        else:
+            contents = bundle.read()
+            bundle_size = len(contents)
+            bundle_hash = hashlib.md5(contents).hexdigest()
+
+            prepare_deploy_result = self.client.prepare_deploy(
+                app_id,
+                deployment_name,
+                bundle_size,
+                bundle_hash,
+            )
+
+            upload_url = prepare_deploy_result.presigned_url
+            parsed_upload_url = urlparse(upload_url)
+            with S3Client(
+                "{}://{}".format(parsed_upload_url.scheme, parsed_upload_url.netloc), timeout=120
+            ) as s3_client:
+                upload_result = s3_client.upload(
+                    "{}?{}".format(parsed_upload_url.path, parsed_upload_url.query),
+                    prepare_deploy_result.presigned_checksum,
+                    bundle_size,
+                    contents,
+                )
+                S3Server(upload_url).handle_bad_response(upload_result)
+
+            self.client.do_deploy(prepare_deploy_result.bundle_id, prepare_deploy_result.app_id)
```

Comment:
I don't know the best place to inject this. But the task that does the deploy can fail. In that case it probably doesn't make sense to send the user to the `app_url` (because the app_url would either not work or not be the version of the app they think they just uploaded).

So I might check for this in `wait_until_task_is_successful`... and print and error and exception? I think that's what the connect side does (https://github.com/rstudio/rsconnect-python/blob/master/rsconnect/api.py#L45-L52).

Reply:
Updated to handle error/failed status as an exception.